File: backend/sdk/loaders/guidelines.py
"""
Guidelines and system prompt configuration.

Provides functions for loading the base system prompt from guidelines.yaml.
"""

import logging

from .yaml_loaders import get_guidelines_config

logger = logging.getLogger(__name__)


def get_base_system_prompt(agent_name: str | None = None) -> str:
    """
    Load the base system prompt from guidelines.yaml.

    Selects system prompt based on agent type:
    - Action Manager: Uses 'system_prompt_AM' if available
    - Onboarding Manager: Uses 'system_prompt_OM' if available
    - Other agents: Uses 'active_system_prompt' field (default behavior)

    Character configuration is always appended to the system prompt with markdown headings.

    Args:
        agent_name: Optional agent name to determine which prompt variant to use.
                    If None, uses the default active_system_prompt.

    Returns:
        The system prompt template with {agent_name} placeholder
    """
    from core.settings import DEFAULT_FALLBACK_PROMPT
    from domain.entities.agent import is_action_manager, is_onboarding_manager

    try:
        guidelines_config = get_guidelines_config()

        # Determine which system prompt to use based on agent type
        if agent_name:
            if is_action_manager(agent_name):
                # Action Manager: prefer system_prompt_AM
                system_prompt = guidelines_config.get("system_prompt_AM", "")
                if system_prompt:
                    logger.debug(f"Using system_prompt_AM for agent '{agent_name}'")
                    return system_prompt.strip()
            elif is_onboarding_manager(agent_name):
                # Onboarding Manager: prefer system_prompt_OM
                system_prompt = guidelines_config.get("system_prompt_OM", "")
                if system_prompt:
                    logger.debug(f"Using system_prompt_OM for agent '{agent_name}'")
                    return system_prompt.strip()

        # Default: use active_system_prompt selector
        # Falls back to "system_prompt" if not specified
        active_prompt_key = guidelines_config.get("active_system_prompt", "system_prompt")
        system_prompt = guidelines_config.get(active_prompt_key, "")

        # If active key not found, try default "system_prompt"
        if not system_prompt and active_prompt_key != "system_prompt":
            logger.warning(f"System prompt '{active_prompt_key}' not found, falling back to 'system_prompt'")
            system_prompt = guidelines_config.get("system_prompt", "")

        if system_prompt:
            return system_prompt.strip()
        else:
            logger.warning("system_prompt not found in guidelines.yaml, using fallback")
            return DEFAULT_FALLBACK_PROMPT
    except Exception as e:
        # Log and use fallback on any error
        logger.error(f"Error loading system prompt from guidelines.yaml: {e}")
        return DEFAULT_FALLBACK_PROMPT

File: backend/src/sdk/loaders/guidelines.ts
// `guidelines_3rd.yaml` holds several prompt bodies plus an
// `active_system_prompt` key naming which one is live, so a prompt is swapped
// by editing one line. The Action and Onboarding Managers opt out.

import { DEFAULT_FALLBACK_PROMPT } from '@/config/settings'
import { isActionManager, isOnboardingManager } from '@/domain/agent'
import { getGuidelinesConfig } from './yaml-config'
import { getLogger } from '@/infrastructure/logging/logger'

const logger = getLogger('Guidelines')

// Defined in `domain/agent.ts`; re-exported because callers import them here.
export { isActionManager, isOnboardingManager }

function readPrompt(config: Record<string, unknown>, key: string): string {
  const value = config[key]
  return typeof value === 'string' ? value : ''
}

/**
 * `system_prompt_AM` for the Action Manager, `system_prompt_OM` for the
 * Onboarding Manager, otherwise `active_system_prompt`'s key. The result still
 * holds `{agent_name}` placeholders.
 */
export function getBaseSystemPrompt(agentName?: string | null): string {
  try {
    const config = getGuidelinesConfig()

    if (agentName) {
      if (isActionManager(agentName)) {
        const prompt = readPrompt(config, 'system_prompt_AM')
        if (prompt) return prompt.trim()
      } else if (isOnboardingManager(agentName)) {
        const prompt = readPrompt(config, 'system_prompt_OM')
        if (prompt) return prompt.trim()
      }
      // A missing AM/OM prompt intentionally falls through to the selector
      // rather than erroring, so deleting the key degrades to the shared prompt.
    }

    const activeKey = readPrompt(config, 'active_system_prompt') || 'system_prompt'
    let prompt = readPrompt(config, activeKey)

    if (!prompt && activeKey !== 'system_prompt') {
      logger.warning(`System prompt '${activeKey}' not found, falling back to 'system_prompt'`)
      prompt = readPrompt(config, 'system_prompt')
    }

    if (prompt) return prompt.trim()

    logger.warning('system_prompt not found in guidelines config, using fallback')
    return DEFAULT_FALLBACK_PROMPT
  } catch (error) {
    logger.error(`Error loading system prompt: ${String(error)}`)
    return DEFAULT_FALLBACK_PROMPT
  }
}
